userspace/libsinsp/container_engine/containerd.cpp

```diff
@@ -0,0 +1,231 @@
+		        container.m_id.c_str(),
+		        status.error_message().c_str());
+		return false;
+	}
+
+	auto containers = resp.containers();
+
+	if(containers.size() == 0) {
+		libsinsp_logger()->format(sinsp_logger::SEV_DEBUG,
+		                          "containerd (%s): ListContainerResponse status error message: "
+		                          "(container id has no match)",
+		                          container.m_id.c_str());
+		return false;
+	} else if(containers.size() > 1) {
+		libsinsp_logger()->format(sinsp_logger::SEV_DEBUG,
+		                          "containerd (%s): ListContainerResponse status error message: "
+		                          "(container id has more than one match)",
+		                          container.m_id.c_str());
+		return false;
+	}
+
+	// Usually the image has this form: `docker.io/library/ubuntu:22.04`
+	auto raw_image_splits = sinsp_split(containers[0].image(), ':');
+
+	container.m_id = container_id;
+	container.m_full_id = containers[0].id();
+	// We assume that the last `/`-separated field is the image
+	container.m_image = raw_image_splits[0].substr(raw_image_splits[0].rfind("/") + 1);
+	// and the first part is the repo
+	container.m_imagerepo = raw_image_splits[0].substr(0, raw_image_splits[0].rfind("/"));
+	container.m_imagetag = raw_image_splits[1];
+	container.m_imagedigest = "";
+	container.m_type = CT_CONTAINERD;
+
+	// Retrieve the labels.
+	for(const auto &pair : containers[0].labels()) {
+		if(pair.second.length() <= sinsp_container_info::m_container_label_max_length) {
+			container.m_labels[pair.first] = pair.second;
+		}
+	}
+
+	// The spec field keeps the information about the mounts.
+	Json::Value spec;
+	Json::Reader reader;
+	// The spec field of the response is just a raw json.
+	reader.parse(containers[0].spec().value(), spec);
+
+	// Retrieve the mounts.
+	for(const auto &m : spec["mounts"]) {
+		bool readonly = false;
+		std::string mode;
+		for(const auto &jopt : m["options"]) {
+			std::string opt = jopt.asString();
+			if(opt == "ro") {
+				readonly = true;
+			} else if(opt.rfind("mode=") == 0) {
+				mode = opt.substr(5);
+			}
+		}
+		container.m_mounts.emplace_back(m["source"].asString(),
+		                                m["destination"].asString(),
+		                                mode,
+		                                !readonly,
+		                                spec["linux"]["rootfsPropagation"].asString());
+	}
+
+	// Retrieve the env.
+	for(const auto &env : spec["process"]["env"]) {
+		container.m_env.emplace_back(env.asString());
+	}
+
+	return true;
+}
+
+bool libsinsp::container_engine::containerd::resolve(sinsp_threadinfo *tinfo,
+                                                     bool query_os_for_missing_info) {
+	auto container = sinsp_container_info();
+	std::string container_id, cgroup;
+
+	if(!matches_runc_cgroups(tinfo, CONTAINERD_CGROUP_LAYOUT, container_id, cgroup)) {
+		return false;
+	}
+
+	if(!parse_containerd(container, container_id)) {
```

Comment:
Shouldn't the API call be async?

Reply:
Oh yes! I removed that part during a debugging session. I'll readd it.